File: grpc/grpc-client-annotation-processor/src/main/java/ru/tinkoff/grpc/client/annotation/processor/GrpcClientExtensionFactory.java
package ru.tinkoff.grpc.client.annotation.processor;

import ru.tinkoff.kora.kora.app.annotation.processor.extension.ExtensionFactory;
import ru.tinkoff.kora.kora.app.annotation.processor.extension.KoraExtension;

import javax.annotation.processing.ProcessingEnvironment;
import java.util.Optional;

import static ru.tinkoff.grpc.client.annotation.processor.GrpcClassNames.abstractStub;

public class GrpcClientExtensionFactory implements ExtensionFactory {
    @Override
    public Optional<KoraExtension> create(ProcessingEnvironment processingEnvironment) {
        var stubTypeElement = processingEnvironment.getElementUtils().getTypeElement(abstractStub.canonicalName());
        if (stubTypeElement == null) {
            return Optional.empty();
        }
        var stubErasure = processingEnvironment.getTypeUtils().erasure(stubTypeElement.asType());
        return Optional.of(new GrpcClientExtension(processingEnvironment, stubErasure));
    }
}

File: grpc/grpc-client-symbol-processor/src/main/kotlin/ru.tinkoff.grpc.client.ksp/GrpcClientExtensionFactory.kt
package ru.tinkoff.grpc.client.ksp

import com.google.devtools.ksp.processing.CodeGenerator
import com.google.devtools.ksp.processing.KSPLogger
import com.google.devtools.ksp.processing.Resolver
import ru.tinkoff.grpc.client.ksp.GrpcClassNames.abstractCoroutineStub
import ru.tinkoff.grpc.client.ksp.GrpcClassNames.abstractStub
import ru.tinkoff.kora.kora.app.ksp.extension.ExtensionFactory
import ru.tinkoff.kora.kora.app.ksp.extension.KoraExtension
import ru.tinkoff.kora.ksp.common.KspCommonUtils.getClassDeclarationByName

class GrpcClientExtensionFactory : ExtensionFactory {
    override fun create(resolver: Resolver, kspLogger: KSPLogger, codeGenerator: CodeGenerator): KoraExtension? {
        val abstractStub = resolver.getClassDeclarationByName(abstractStub)
        val abstractCoroutineStub = resolver.getClassDeclarationByName(abstractCoroutineStub)
        if (abstractStub == null && abstractCoroutineStub == null) {
            return null
        }
        return GrpcClientExtension(resolver, kspLogger, codeGenerator, abstractStub, abstractCoroutineStub)
    }
}
